pallene/reaching_definitions.lua

```diff
@@ -0,0 +1,497 @@
+    return new
+end
+-- creates a new node, should only be called to non-control commands.
+-- returns an updated context
+function CFG:new_node(cmd,ctx)
+    table.insert(self,{
+        cmd = cmd, --TODO should be a list of commands
+        to={},
+        from={}
+    })
+    local new_node_id = #self
+    local new_node = self[new_node_id]
+    Set.copy(new_node.from, ctx.current_node)
+    for _,from in ipairs(new_node.from) do
+            table.insert(self[from].to,new_node_id)
+    end
+    local newctx = CFG.new_ctx(ctx)
+    newctx.current_node = {new_node_id}
+    return newctx
+end
+-- creates a new edge from each of the current nodes to each of the target nodes
+function CFG:new_edge(target,ctx)
+    for _,from in ipairs(ctx.current_node) do
+        for _,to in ipairs(target) do
+            table.insert(self[from].to,to)
+            table.insert(self[to].from,from)
+        end
+    end
+    return ctx
+end
+
+-- traverses the CFG applying the array of functions fs with the input to each node.
+function CFG:traverse(index,fs,input,visited)
+    visited = visited or {}
+    visited[index] = true
+    for _,v in ipairs(fs) do v(self,index,input) end
+    assert(self[index], 'failed to retrieve node at index '..index)
+    for _,v in ipairs(self[index].to) do
+        -- dont go to entry or exit nodes
+        if v > 0 and not visited[v] then
+            self:traverse(v,fs,input,visited)
+        end
+    end
+end
+-- a dead code analysis at CFG level:
+-- If a node's from is empty, 
+-- then the node isn't reachable.
+-- returns a set of localizations of unreachable code in the ret input.
+-- if it returns ret[true] as unreachable, 
+-- it means that the exit node is unreachable,
+-- meaning that maybe there is an inifinite loop in the function.
+function CFG:check_unreachable(index, ret)
+    local node = self[index]
+    if node then
+        if #node.from == 0 then
+            if node.cmd.loc then
+                ret[node.cmd.loc] = true
+            else
+                error('impossible',node._tag)
+            end 
+        end
+    end
+end
+function CFG:unreachable()
+    local ret = {}
+    self:traverse(1,{CFG.check_unreachable},ret)
+    if #self[-1].from ~= 0 then
+        ret[true] = true
+    end
+    return ret
+end
+-- converts the CFG to DOT format, which can be redered with the graphviz utility.
+function CFG:to_dot()
+    print(' digraph cfg {')
+    print('0 [label="Entry"];')
+    print('-1 [label="Exit"];')
+    local visited = {}
+    self:traverse(1,{function(index,cfg_)
+        local n = cfg_[index]
+        if n and not visited[n] then
+            visited[n] = true
+            print(index..' [label="'..n.cmd._tag..'"];')
+            for _,f in ipairs(n.from) do 
+                print(f..'->'..index..';')
+            end
+        end
+    end})
+    print('}')
+end
+--]]
+-- end   control flow -------------
+-- start definitions  -------------
+
+local function find_defs_uses(cfg,index,i)
```

Comment:
Please do not use the name `i` for things that are not array indexes.

Reply:
it meant input, but ok, will change that